### PR TITLE
Canonicalizers and Declarative Rewrite Patterns

### PR DESCRIPTION
**Pull request by j2kun**

### CODE REVIEW

File: lib/Dialect/Poly/PolyPatterns.td
Comment (UtkarshKunwar) [inline lib/Dialect/Poly/PolyPatterns.td:1]: I was following the tutorial and got stuck here with MLIR v18.1.x. I've described the issue on the [LLVM repo here in detail](https://github.com/llvm/llvm-project/issues/92227). Just wanted to know if someone else has also faced this issue as I couldn't find the solution for it.

Comment (j2kun) [reply lib/Dialect/Poly/PolyPatterns.td:1]: Interesting. I haven't seen people use MLIR by installing the tools as separate binaries, and instead they tend to build everything from source at a specific LLVM commit (or HEAD).

Comment (UtkarshKunwar) [reply lib/Dialect/Poly/PolyPatterns.td:1]: I must admit I'm not very familiar with the standard MLIR development process. Building a large project like LLVM can take some time on potato computers or in CI/CD setups. So I prefer stable binary releases for such dependencies as they enable faster "clean build" iterations. The dependencies are supposed to be frozen anyway! :)

I also faced problems with some MLIR support tools (LSP servers in particular) when building from source (I filed that issue [here](https://github.com/llvm/llvm-project/issues/91263)).

Comment (mtsokol) [reply lib/Dialect/Poly/PolyPatterns.td:1]: FWIW We build LLVM from source in the GitHub CI, and I also think that "from source" LLVM build is necessary to build an out-of-tree dialect.
With `ccache` for the compilation caching, the whole LLVM+MLIR+minimal dialect build process takes about 18min on the free macos-14 M1 runner [link](https://github.com/finch-tensor/Finch-mlir/actions/runs/12744124849/job/35515428383).

**Pull request merged**